topiary-core/src/atom_collection.rs

```diff
@@ -883,13 +896,62 @@ impl AtomCollection {
         }
     }
 
+    /// Separate post processing of capitalisation, to avoid confusion around whitespacing.
+    fn post_process_capitalization(&mut self) {
+        let mut case_context: Vec<Capitalisation> = Vec::new();
+        for atom in &mut self.atoms {
+            match atom {
+                Atom::CaseBegin(case) => match case {
+                    Capitalisation::UpperCase => {
+                        case_context.push(Capitalisation::UpperCase);
+                        *atom = Atom::Empty;
+                    }
+                    Capitalisation::LowerCase => {
+                        case_context.push(Capitalisation::LowerCase);
+                        *atom = Atom::Empty;
+                    }
+                    Capitalisation::Pass => {
+                        case_context.push(Capitalisation::Pass);
+                        *atom = Atom::Empty;
+                    }
+                },
+                Atom::CaseEnd => {
+                    case_context.pop();
+                    *atom = Atom::Empty;
+                }
+                Atom::Leaf {
+                    content,
+                    id,
+                    original_position,
+                    single_line_no_indent,
+                    multi_line_indent_all,
+                    ..
+                } => {
+                    *atom = Atom::Leaf {
+                        content: (*content).to_string(),
+                        id: *id,
+                        original_position: *original_position,
+                        single_line_no_indent: *single_line_no_indent,
+                        multi_line_indent_all: *multi_line_indent_all,
+                        capitalisation: case_context
+                            .last()
+                            .unwrap_or(&Capitalisation::Pass)
+                            .clone(),
+                    }
+                }
```

Comment:
@yannham and I both noticed that what you're effectively doing here is an in-place mutation of the `capitalisation` element of `atom` (which is an `&mut Atom`). As such, this is probably a little easier to read:

```suggestion
                Atom::Leaf { capitalisation, .. } => {
                    *capitalisation = case_context.last().unwrap_or(&Capitalisation::Pass).clone();
                }
```